src/process_quantiles.R

```diff
@@ -0,0 +1,184 @@
+  hru_id_end <- hru_id_start + (n_hrus_per_task-1) # which hru to end with
+  
+  # If we are past the last column, cut the final col id off
+  n_hrus <- 109951
+  hru_id_end <- ifelse(hru_id_end > n_hrus,
+                       yes = n_hrus,
+                       no = hru_id_end)
+  
+  hru_seq <- hru_id_start:hru_id_end
+  return(hru_seq)
+}
+read_ncdf_data <- function(fn, varid, hru_seq) {
+  
+  # The file is NetCDF
+  nc <- nc_open(fn)
+  
+  message(sprintf("Reading in NetCDF data for %s", varid))
+  # Only load rows for current HRUs
+  data_nc <- ncvar_get(nc, varid, 
+                       start = c(head(hru_seq, 1), 1),
+                       count = c(length(hru_seq), -1))
+  time <- ncvar_get(nc, "time")
+  hruid <- ncvar_get(nc, "hruid")[hru_seq]
+  
+  # Convert ncdf4 times to R dates
+  time_att <- ncdf4::ncatt_get(nc, "time")
+  time_start <- as.Date(gsub("days since ", "", time_att$units))
+  time_fixed <- time + time_start # creates dates from the "days since" var
+  
+  # Keep only complete years of data
+  # data_years <- format(time_fixed, "%Y")
+  # n_days_per_year <- table(data_years)
+  # data_years_complete <- names(n_days_per_year)[n_days_per_year >= 365]
+  
+  year_doy_vector <- paste(format(time_fixed, "%Y"), format(time_fixed, "%j"), sep = "_")
+  dt <- as.data.table(cbind(hruid, data_nc)) # takes ~2 minutes
+  names(dt) <- c("hruid", year_doy_vector)
+  dt_long <- melt(dt, id.vars = "hruid", variable.name = "year_doy")
+  dt_long[, variable := varid]
+  
+  return(dt_long)
+}
+get_doy_sequence <- function(target_doy, window = 5) {
+  
+  # Determine which doy to start & end with in the window
+  begin_doy <- target_doy - window
+  end_doy <- target_doy + window
+  
+  # Handle edge cases:
+  last_doy <- 365 # switch back to 366 once leap days are implemented
+  if(begin_doy <= 0) {
+    # Need to look backward into last days of the year
+    seq_doy <- c((last_doy + begin_doy):last_doy, 1:end_doy)
+  } else if (end_doy > last_doy) {
+    # Need to look forward into first days of the year
+    seq_doy <- c(begin_doy:last_doy, 1:(end_doy - last_doy))
+  } else {
+    seq_doy <- begin_doy:end_doy
+  }
+  
+  return(seq_doy)
+}
+extract_doy_cols_to_vec <- function(df, doy_seq) {
+  ############## WORKING ON THIS
+  # Do I need to convert to df?
+  df <- as.data.frame(df)
+  doy_seq_colnames <- sprintf("%03d", doy_seq)
+  target_df <- df[, doy_seq_colnames]
+  ############## 
+  target_vector <- unlist(target_df, use.names = FALSE)
+  return(target_vector)
+}
+get_quantile_df <- function(target_doy_values, target_doy_seq, 
+                            target_doy, probs = c(0.10, 0.25, 0.75, 0.90)) {
+  target_doy_quantiles <- quantile(target_doy_values, 
+                                   probs = probs, 
+                                   type = 6,
+                                   na.rm = TRUE)
+  target_doy_quantile_df <- cbind(DOY = target_doy, t(target_doy_quantiles))
```

Comment:
I think you get a tiny but real performance bump by returning
`c(target_doy, target_doy_quantiles)`
instead of this matrix row. The tradeoff is that you do have to add the probs back to your matrix colnames later.